src/client/datascience/notebook/kernelProvider.ts

```diff
@@ -202,146 +114,79 @@ export class VSCodeKernelPickerProvider implements INotebookKernelProvider {
+                        r.kernelSpec.metadata?.vscode?.extension_id &&
+                        this.extensions.getExtension(r.kernelSpec.metadata?.vscode?.extension_id)
+                    ) {
+                        return false;
                     }
-                    return sessionManager;
-                },
-                token
-            );
-        }
-    }
-    private async getJupyterSessionManager(resource: Uri) {
-        if (this.isLocalLaunch) {
-            return;
-        }
-        try {
-            // Make sure we have a connection or we can't get remote kernels.
+                }
+                return true;
+            });
+        } else {
             const connection = await this.notebookProvider.connect({
                 getOnly: false,
-                resource,
+                resource: document.uri,
                 disableUI: false,
                 localOnly: false
             });
-            if (!connection) {
-                throw new Error('Using remote connection but connection is undefined');
-            } else if (connection?.type === 'raw') {
-                throw new Error('Using remote connection but connection type is raw');
-            } else {
-                return this.jupyterSessionManagerFactory.create(connection);
-            }
-        } catch (ex) {
-            // This condition is met when remote Uri is invalid.
-            // User cannot even run a cell, as kernel list is invalid (we can't get it).
-            sendKernelTelemetryEvent(resource, Telemetry.NotebookStart, undefined, undefined, ex);
-            throw ex;
+
+            kernels = await this.remoteKernelFinder.listKernels(document.uri, connection);
+            preferred = await this.remoteKernelFinder.findKernel(
+                document.uri,
+                connection,
+                getNotebookMetadata(document),
+                token
+            );
         }
-    }
-    private createNotebookKernelMetadataFromPreferredKernel(
-        preferredKernel?: KernelConnectionMetadata
-    ): VSCodeNotebookKernelMetadata | undefined {
-        if (!preferredKernel) {
-            return;
-        } else if (preferredKernel.kind === 'startUsingDefaultKernel') {
-            return;
-        } else if (preferredKernel.kind === 'startUsingPythonInterpreter') {
-            return new VSCodeNotebookKernelMetadata(
-                preferredKernel.interpreter.displayName || preferredKernel.interpreter.path,
-                '',
-                preferredKernel.interpreter.path,
-                preferredKernel,
-                true,
-                this.kernelProvider,
-                this.notebook,
-                this.context,
-                this.preferredRemoteKernelIdProvider,
-                this.commandManager
+
+        // If we already have a kernel selected, then set that one as preferred
+        const editor =
+            this.notebook.notebookEditors.find((e) => e.document === document) ||
```

Comment:
**Suggestion - I would suggest using the following two options:**
Absolutely not necessary, but a good place to discuss perf issues so you too are aware of this.

* Check property `activeNotebookEditor.kernel`
	* Exsiting code
* If this is not available, then create a private weakmap and use document as the key.
	* This is for scenarios where user opens a notebok & opens another notebook
	* Then go back to previous notebook, VSC will NOT set notebookEditor.kernel property
	* I.e. if you open notebook & toggle between notebooks or other files, VS Code will ALWAYS keep calling this.
	* We could improve the perf by caching ourselves against the document.
* I found out on friday that VS Code only sets the Kernel property against Editor only when user runs a cell or edits a notebook.